From: Filipe Manana <fdmanana@suse.com>
Date: Thu, 4 Aug 2016 11:40:29 +0100
Patch-mainline: Not yet, used for debugging issues in openqa only
References: bsc#990384
Subject: [PATCH] Btrfs: log error when an extent item is not found for update

When running delayed references and we need to add a new backreference for
an extent, if we don't find the extent item we end up aborting the current
transaction with an EIO error, leaving the filesystem in readonly mode,
because this is a critical inconsistent/corrupt metadata state. So when
this happens log details about the extent and the new reference to help
post-mortem analysis and debugging.

Signed-off-by: Filipe Manana <fdmanana@suse.com>
---
 fs/btrfs/extent-tree.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/fs/btrfs/extent-tree.c b/fs/btrfs/extent-tree.c
index 47cdc6f..20b400f 100644
--- a/fs/btrfs/extent-tree.c
+++ b/fs/btrfs/extent-tree.c
@@ -1607,6 +1607,12 @@ again:
 		err = -ENOENT;
 		goto out;
 	} else if (WARN_ON(ret)) {
+		btrfs_err(root->fs_info,
+			  "did not found extent item for bytenr %llu num_bytes %llu owner %llu skinny metadata %d slot %d",
+			  bytenr, num_bytes, owner,
+			  btrfs_fs_incompat(root->fs_info, SKINNY_METADATA),
+			  path->slots[0]);
+		btrfs_print_leaf(root, path->nodes[0]);
 		err = -EIO;
 		goto out;
 	}
-- 
1.8.4.5

